x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/TrainedModelStatsService.java

```diff
@@ -97,11 +103,26 @@ public void beforeStop() {
         clusterService.addListener((event) -> this.clusterState = event.state());
     }
 
-    public void queueStats(InferenceStats stats) {
+    /**
+     * Queues the stats for storing.
+     * @param stats The stats to store or increment
+     * @param flush When `true`, this indicates that stats should be written as soon as possible.
+     *              If `false`, stats are not persisted until the next periodic persistence action.
+     */
+    public void queueStats(InferenceStats stats, boolean flush) {
+        if (stats.hasStats() == false) {
+            if (flush) {
+                threadPool.executor(MachineLearning.UTILITY_THREAD_POOL_NAME).execute(this::updateStats);
```

Comment:
2 threads could be calling this method at the same time, one invoked here and the other from the the scheduler. Is that safe? Both threads will be iterating and removing entries from the map, maybe use an iterator on the entrySet

Reply:
It should be safe. I make a copy of the entry set and iterate removals. If the key does not exist, it would return `null`. So, only one of the two threads would end up having the update.

Reply:
More verbose reply. 

The iteration code is here:
```
Set<String> keys = new HashSet<>(statsQueue.keySet());
for(String k : keys) {
    InferenceStats inferenceStats = statsQueue.remove(k);
    if (inferenceStats != null) {
        stats.add(inferenceStats);
    }
}
```
This gets the latest known view of the keySet and makes a copy. This way if keys are added AFTER the remove, it does not double count the stats (I think the keySet view is updated when keys are added...so I opted to not use it directly in iteration).

concurrent map `remove` is thread safe. So if one thread removes an item from the hashmap, the other thread won't see it and get `null`. So, it won't add it to its request payload.

Reply:
👍 it was the double removal I was worried about